### PR TITLE
e2e: test snapshot restore

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/6391.

### CODE REVIEW

**Comment by gyuho** [comment]
Considered refactoring e2e code base but it requires much more changes, in my opinion.

/cc @xiang90 @heyitsanthony

File: e2e/ctl_v3_snapshot_test.go [hunk 1]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
 	}
 	return resp, nil
 }
+
+func TestCtlV3SnapshotRestore(t *testing.T) {
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // Issue6361 tests the scenarios:
// (describes what this test checks) 
TestIssue6361

File: e2e/ctl_v3_snapshot_test.go [hunk 2]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+		clusterSize:  1,
+		initialToken: "new",
+		keepDataDir:  true,
+	})
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+
+	defer func() {
+		os.Unsetenv("ETCDCTL_API")
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: move this near line133

File: e2e/ctl_v3_snapshot_test.go [hunk 3]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	// write some keys
+	kvs := []kv{{"foo1", "val1"}, {"foo2", "val2"}, {"foo3", "val3"}}
+	for i := range kvs {
+		if err = spawnWithExpect(append(prefixArgs, "put", kvs[i].key, kvs[i].val), "OK"); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	fpath := filepath.Join(os.TempDir(), "test.snapshot")
+	os.RemoveAll(fpath)
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: should we remove line163?

File: e2e/ctl_v3_snapshot_test.go [hunk 4]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	defer os.RemoveAll(fpath)
+
+	// save snapshot
+	if err = spawnWithExpect(
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: let's make this one line

```go
err = spawnWithExpect(...)
if err != nil {
    ...
}
```

File: e2e/ctl_v3_snapshot_test.go [hunk 5]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+		append(prefixArgs, "snapshot", "save", fpath),
+		fmt.Sprintf("Snapshot saved at %s", fpath),
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = epc.backends()[0].Stop(); err != nil {
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: we should rename backends to process or member. backends is confusing. we can do this in another pr though.

File: e2e/ctl_v3_snapshot_test.go [hunk 6]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	var newDataDir string
+	newDataDir, err = ioutil.TempDir("", "newdata")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.RemoveAll(newDataDir)
+	defer os.RemoveAll(newDataDir)
+
+	// restore the snapshot
+	if err = spawnWithExpect(
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: same as the comment for line167

File: e2e/ctl_v3_snapshot_test.go [hunk 7]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	var newDataDir2 string
+	newDataDir2, err = ioutil.TempDir("", "newdata2")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.RemoveAll(newDataDir2)
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: remove this line?

File: e2e/etcd_test.go [hunk 8]
```diff
@@ -140,6 +140,10 @@ type etcdProcessConfig struct {
 	acurltls  string
 	acurlHost string
 
+	initialToken   string
+	name           string
```
Comment (xiang90) [inline e2e/etcd_test.go]: move name above purl

File: e2e/ctl_v3_snapshot_test.go [hunk 9]
```diff
@@ -119,3 +126,152 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+			t.Fatal(err)
+		}
+	}
+
+	if err = nepc.Stop(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func getHost(ep string) string {
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: do we really need this? etcdctl can accept http:// i believe.

**Comment by xiang90** [comment]
@gyuho looks good in general. Let's clean this up. I will do another round of review.

File: e2e/ctl_v3_snapshot_test.go (hunk 5 above)
Comment (gyuho) [inline e2e/ctl_v3_snapshot_test.go]: Yeah it's confusing. Will change in a separate PR. Thanks.

**Comment by gyuho** [comment]
@xiang90 All addressed. PTAL. Thanks!

File: e2e/ctl_v3_snapshot_test.go [hunk 10]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	if err = epc.backends()[0].Stop(); err != nil {
+		t.Fatal(err)
+	}
+
+	var newDataDir string
+	newDataDir, err = ioutil.TempDir("", "newdata")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.RemoveAll(newDataDir) // delete temp dir created by ioutil.TempDir
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: shall we remove this line? this is a dup with line 180. no?

File: e2e/ctl_v3_snapshot_test.go [hunk 11]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	defer os.RemoveAll(newDataDir)
+
+	// restore the snapshot
+	if err = spawnWithExpect([]string{ctlBinPath, "snapshot", "restore", fpath, "--name", epc.procs[0].cfg.name, "--initial-cluster", epc.procs[0].cfg.initialCluster, "--initial-cluster-token", epc.procs[0].cfg.initialToken, "--initial-advertise-peer-urls", epc.procs[0].cfg.purl.String(), "--data-dir", newDataDir}, "membership: added member"); err != nil {
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]:
```go

err = xxx (multiple lines)
if err!= nil {
    ....
}
```

File: e2e/ctl_v3_snapshot_test.go [hunk 12]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	}
+
+	// ensure restoration from snapshot has the correct data
+	for i := range kvs {
+		if err = spawnWithExpect(append(prefixArgs, "get", kvs[i].key), kvs[i].val); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	clientURL := fmt.Sprintf("http://localhost:%d", etcdProcessBasePort+21)
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: why + 21?

File: e2e/ctl_v3_snapshot_test.go [hunk 13]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	if err = spawnWithExpect(append(prefixArgs, "member", "add", "newmember", fmt.Sprintf("--peer-urls=%s", peerURL)), " added to cluster "); err != nil {
+		t.Fatal(err)
+	}
+
+	var newDataDir2 string
+	newDataDir2, err = ioutil.TempDir("", "newdata2")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.RemoveAll(newDataDir2) // delete temp dir created by ioutil.TempDir
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: this is dup with 217, no?

File: e2e/ctl_v3_snapshot_test.go [hunk 14]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	for i := range kvs {
+		if err = spawnWithExpect(append(prefixArgs, "put", kvs[i].key, kvs[i].val), "OK"); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	fpath := filepath.Join(os.TempDir(), "test.snapshot")
+	defer os.RemoveAll(fpath)
+
+	// save snapshot
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // etcdctl save snapshot

File: e2e/ctl_v3_snapshot_test.go [hunk 15]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+
+	var newDataDir string
+	newDataDir, err = ioutil.TempDir("", "newdata")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.RemoveAll(newDataDir) // delete temp dir created by ioutil.TempDir
+	defer os.RemoveAll(newDataDir)
+
+	// restore the snapshot
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // etcdctl restore the snapshot

File: e2e/ctl_v3_snapshot_test.go [hunk 16]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	if err = spawnWithExpect([]string{ctlBinPath, "snapshot", "restore", fpath, "--name", epc.procs[0].cfg.name, "--initial-cluster", epc.procs[0].cfg.initialCluster, "--initial-cluster-token", epc.procs[0].cfg.initialToken, "--initial-advertise-peer-urls", epc.procs[0].cfg.purl.String(), "--data-dir", newDataDir}, "membership: added member"); err != nil {
+		t.Fatal(err)
+	}
+
+	// start a node from the restored snapshot
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: start the etcd member using the restored snapshot

File: e2e/ctl_v3_snapshot_test.go [hunk 17]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	for i := range epc.procs[0].cfg.args {
+		if epc.procs[0].cfg.args[i] == "--data-dir" {
+			epc.procs[0].cfg.args[i+1] = newDataDir
+		}
+	}
+	if err = epc.backends()[0].Restart(); err != nil {
+		t.Fatal(err)
+	}
+
+	// ensure restoration from snapshot has the correct data
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // ensure the restored member has the correct data

File: e2e/ctl_v3_snapshot_test.go [hunk 18]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	for i := range kvs {
+		if err = spawnWithExpect(append(prefixArgs, "get", kvs[i].key), kvs[i].val); err != nil {
+			t.Fatal(err)
+		}
+	}
+
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // add a new member into the cluster

File: e2e/ctl_v3_snapshot_test.go [hunk 19]
```diff
@@ -119,3 +125,122 @@ func getSnapshotStatus(cx ctlCtx, fpath string) (snapshotStatus, error) {
...
+	clientURL := fmt.Sprintf("http://localhost:%d", etcdProcessBasePort+21)
+	peerURL := fmt.Sprintf("http://localhost:%d", etcdProcessBasePort+31)
+	if err = spawnWithExpect(append(prefixArgs, "member", "add", "newmember", fmt.Sprintf("--peer-urls=%s", peerURL)), " added to cluster "); err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (xiang90) [inline e2e/ctl_v3_snapshot_test.go]: // start the new member

**Comment by gyuho** [comment]
@xiang90 Just addressed.

`os.RemoveAll(newDataDir)` is needed, because `snapshot restore` command requires the `--data-dir` to be not-exist before it runs the command (https://github.com/coreos/etcd/blob/master/etcdctl/ctlv3/command/snapshot_command.go#L180-L190)


PTAL. Thanks.

**Comment by xiang90** [comment]
@gyuho OK. Can we just get the temp dir name somehow without creating it then? The code is really confusing as it is.

**Comment by gyuho** [comment]
@xiang90 Fixed. (just using os.TempDir)

**Comment by xiang90** [comment]
@gyuho LGTM. I think you now understand why I want to refactor our e2e infra well. In this pr, we have to use raw processExpect all over the place. We cannot utilize existing code :(. But let's see if this causes us problem in the future.

**Comment by gyuho** [comment]
@xiang90 Yeah this is not the best way, but we can refactor more later. Merging in greens. Thanks.
